icu4c/source/tools/icuexportdata/icuexportdata.cpp

```diff
@@ -96,6 +167,110 @@ void dumpEnumeratedProperty(UProperty uproperty, FILE* f) {
+
+    // Create a mutable UCPTrie builder populated with Script property values data.
+    const UCPMap* scInvMap = u_getIntPropertyMap(UCHAR_SCRIPT, status);
+    handleError(status, scxFullPropName);
+    LocalUMutableCPTriePointer builder(umutablecptrie_fromUCPMap(scInvMap, status));
+    handleError(status, scxFullPropName);
+
+    // The values for the output scx companion array.
+    // Invariant is that all subvectors are distinct.
+    std::vector< std::vector<uint16_t> > outputDedupVec;
+
+    // The sc/scx companion array is an array of arrays (of script codes)
+    fputs("script_code_array = [\n", f);
+    for(const UChar32 cp : scxCodePoints) {
+        // Get the Script value
+        uint32_t scVal = umutablecptrie_get(builder.getAlias(), cp);
+        // Get the Script_Extensions value (array of Script codes)
+        const int32_t SCX_ARRAY_CAPACITY = 32;
+        UScriptCode scxValArray[SCX_ARRAY_CAPACITY];
+        int32_t numScripts = uscript_getScriptExtensions(cp, scxValArray, SCX_ARRAY_CAPACITY, status);
+        handleError(status, scxFullPropName);
+
+        // Convert the scx array into a vector
+        std::vector<uint16_t> scxValVec;
+        for(int i = 0; i < numScripts; i++) {
+            scxValVec.push_back(scxValArray[i]);
+        }
+        // Ensure that it is sorted
+        std::sort(scxValVec.begin(), scxValVec.end());
+        // Copy the Script value into the first position of the scx array only
+        // if we have the "other" case (Script value is not Common nor Inherited).
+        // This offers faster access when users want only the Script value.
+        if (scVal != USCRIPT_COMMON && scVal != USCRIPT_INHERITED) {
+            scxValVec.insert(scxValVec.begin(), scVal);
+        }
+
+        // See if there is already an scx value array matching the newly built one.
+        // If there is, then use its index.
+        // If not, then append the new value array.
+        bool isScxValUnique = true;
+        size_t outputIndex = 0;
+        for (outputIndex = 0; outputIndex < outputDedupVec.size(); outputIndex++) {
+            if (outputDedupVec[outputIndex] == scxValVec) {
+                isScxValUnique = false;
+                break;
+            }
+        }
+
+        if (isScxValUnique) {   
+            outputDedupVec.push_back(scxValVec);
+            usrc_writeArray(f, "  [", scxValVec.data(), 16, scxValVec.size(), "    ", "],\n");
+        }
+
+        // We must update the value in the UCPTrie for the code point to contain:
+        // 9..0 the Script code in the lower 10 bits when 11..10 is 0, else it is
+        //   the index into the companion array 
+        // 11..10 the same higher-order 2 bits in the trie in uprops.icu indicating whether
+        //   3: other
+        //   2: Script=Inherited
+        //   1: Script=Common
+        //   0: Script=value in 9..0 (N/A because we are in this loop to create the companion array for non-0 cases)
+        uint16_t mask = 0;
```

Comment:
Totally optional: You could pull up the mask computation to where you already check for common/inherited and prepend the script code. Slightly less fragile because you would check only once for common/inherited. But fine as is, too.